lib/services/local/sessiontracker.go

```diff
@@ -0,0 +1,379 @@
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sessions := make([]types.SessionTracker, len(sessionList))
+	for i, sessionID := range sessionList {
+		session, err := s.loadSession(ctx, sessionID)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		sessions[i] = session
+	}
+
+	return sessions, nil
+}
+
+// CreateSessionTracker creates a tracker resource for an active session.
+func (s *sessionTracker) CreateSessionTracker(ctx context.Context, req *proto.CreateSessionTrackerRequest) (types.SessionTracker, error) {
+	now := time.Now().UTC()
+
+	spec := types.SessionTrackerSpecV1{
+		SessionID:         req.ID,
+		Kind:              req.Type,
+		State:             types.SessionState_SessionStatePending,
+		Created:           now,
+		Reason:            req.Reason,
+		Invited:           req.Invited,
+		Hostname:          req.Hostname,
+		Address:           req.Address,
+		ClusterName:       req.ClusterName,
+		Login:             req.Login,
+		Participants:      []types.Participant{*req.Initiator},
+		Expires:           req.Expires,
+		KubernetesCluster: req.KubernetesCluster,
+		HostUser:          req.HostUser,
+	}
+
+	session, err := types.NewSessionTracker(spec)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	json, err := marshalSession(session)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	err = s.addSessionToList(ctx, session.GetSessionID())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	item := backend.Item{Key: backend.Key(sessionPrefix, session.GetSessionID()), Value: json}
+	_, err = s.bk.Create(ctx, item)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return session, nil
+}
+
+// UpdateSessionTracker updates a tracker resource for an active session.
+func (s *sessionTracker) UpdateSessionTracker(ctx context.Context, req *proto.UpdateSessionTrackerRequest) error {
+	sessionItem, err := s.bk.Get(ctx, backend.Key(sessionPrefix, req.SessionID))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	session, err := unmarshalSession(sessionItem.Value)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	switch session := session.(type) {
+	case *types.SessionTrackerV1:
+		switch update := req.Update.(type) {
+		case *proto.UpdateSessionTrackerRequest_UpdateState:
+			session.SetState(update.UpdateState.State)
+		case *proto.UpdateSessionTrackerRequest_AddParticipant:
+			session.AddParticipant(*update.AddParticipant.Participant)
+		case *proto.UpdateSessionTrackerRequest_RemoveParticipant:
+			session.RemoveParticipant(update.RemoveParticipant.ParticipantID)
+		}
+	default:
+		return trace.BadParameter("unrecognized session version %T", session)
+	}
+
+	sessionJSON, err := marshalSession(session)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	item := backend.Item{Key: backend.Key(sessionPrefix, req.SessionID), Value: sessionJSON}
+	_, err = s.bk.CompareAndSwap(ctx, *sessionItem, item)
+	if trace.IsCompareFailed(err) {
+		time.Sleep(retryDelay)
+		return s.UpdateSessionTracker(ctx, req)
```

Comment:
Will this recurse forever if it keeps getting compare failed errors?

(Comment applies above as well)